src/game_inventory.cpp

```diff
@@ -92,62 +118,222 @@ int game::inv_for_unequipped( const std::string &title )
     }, _( "You don't have any items to wear." ) );
 }
 
-item_location game::inv_map_splice( item_filter filter, const std::string &title, int radius,
+item_location game::inv_map_splice( const item_filter &filter, const std::string &title, int radius,
                                     const std::string &none_message )
 {
-    return inv_map_splice( convert_filter( filter ), title, radius, none_message );
+    return inv_map_splice( [ &filter ]( const item_location & location ) {
+        return filter( *location.get_item() );
+    }, title, radius, none_message );
 }
 
-item_location game::inv_map_splice( item_location_filter filter, const std::string &title,
+item_location game::inv_map_splice( const item_location_filter &filter, const std::string &title,
                                     int radius,
                                     const std::string &none_message )
 {
-    u.inv.restack( &u );
-    u.inv.sort();
+    return inv_internal( u, inventory_filter_preset( filter ),
+                         title, radius, none_message );
+}
 
-    inventory_pick_selector inv_s( u, title, filter );
+class liquid_inventory_preset: public inventory_selector_preset
+{
+    public:
+        liquid_inventory_preset( const item &liquid ) : liquid( liquid ) {
+            append_cell( [ this ]( const inventory_entry & entry ) {
+                if( is_same_container( entry.get_item() ) ) {
+                    return std::string( _( "is the source" ) );
+                }
+
+                std::string err;
+                const int rem_capacity = get_rem_capacity( entry.get_location(), &err );
+                const int total_capacity = entry.get_item().get_total_capacity_for_liquid( this->liquid );
+
+                if( !err.empty() ) {
+                    return err;
+                }
+
+                return ( entry.get_available_count() > 1 )
+                       ? string_format( "%d x %d/%d", entry.get_available_count(), rem_capacity, total_capacity )
+                       : string_format( "%d/%d", rem_capacity, total_capacity );
+            }, _( "AVAILABLE" ) );
+        }
 
-    inv_s.add_nearby_items( radius );
-    if( inv_s.empty() ) {
-        const std::string msg = ( none_message.empty() ) ? _( "You don't have the necessary item at hand." )
-                                : none_message;
-        popup( msg, PF_GET_KEY );
-        return item_location();
-    }
+        bool is_shown( const item &it ) const override {
+            return it.is_reloadable_with( liquid.typeId() ) || it.is_container();
+        }
 
-    return std::move( inv_s.execute() );
-}
+        bool is_enabled( const item_location &location ) const override {
+            return !is_same_container( *location.get_item() ) && get_rem_capacity( location ) > 0;
+        }
+
+        int get_rank( const item_location &location ) const override {
+            return -get_rem_capacity( location );
```

Comment:
This sort order is rather annoying: it changes a lot and doesn't prioritize filled containers.
Containers with current liquid type should be sorted first.